dhis-2/dhis-services/dhis-service-schema/src/main/java/org/hisp/dhis/schema/introspection/PropertyPropertyIntrospector.java

```diff
@@ -0,0 +1,262 @@
+/*
+ * Copyright (c) 2004-2021, University of Oslo
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ * Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ *
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.schema.introspection;
+
+import static org.hisp.dhis.schema.PropertyType.BOOLEAN;
+import static org.hisp.dhis.schema.PropertyType.CONSTANT;
+import static org.hisp.dhis.schema.PropertyType.DATE;
+import static org.hisp.dhis.schema.PropertyType.REFERENCE;
+import static org.hisp.dhis.system.util.AnnotationUtils.getAnnotation;
+import static org.hisp.dhis.system.util.AnnotationUtils.isAnnotationPresent;
+
+import java.lang.reflect.Method;
+import java.util.Collection;
+import java.util.Date;
+import java.util.EnumSet;
+import java.util.Map;
+import java.util.Set;
+
+import org.hisp.dhis.common.IdentifiableObject;
+import org.hisp.dhis.schema.Property;
+import org.hisp.dhis.schema.PropertyType;
+import org.hisp.dhis.schema.annotation.Property.Access;
+import org.hisp.dhis.schema.annotation.Property.Value;
+import org.hisp.dhis.schema.annotation.PropertyRange;
+import org.hisp.dhis.schema.annotation.PropertyTransformer;
+import org.springframework.util.Assert;
+
+import com.google.common.primitives.Primitives;
+
+/**
+ * A {@link PropertyIntrospector} adds information to existing {@link Property}
```

Comment:
Should it be PropertyPropertyInspector?

Reply:
Yes it is a bit unfortunate that the interface is `PropertyIntrospector` and this implementation is for the `@Property` family of annotations. So it was intentional to have _property_ twice in the name. Any other ideas are very welcome.

Reply:
I got that, I tried to think to a better name without success.
I was just pointing at the name in the Javadoc

```suggestion
 * A {@link PropertyPropertyIntrospector} adds information to existing {@link Property}
```

Reply:
Never mind, I got what you were saying in the Javadoc.

Reply:
Oh, yes the wording is off, I meant to have a _that_ after the class name there.